README.md
This is a plugin to load python dicts representing a domain model into a Neo4j graph

- Scripts handling the import of domain model information from xlxs files or python dictionary files to a neo4j database

## Requirements

- Python2.7
- Extracting data from Excel
  - `xlrd` Version 1.2.0 -> `pip install xlrd`
- Creting cypher queries:
  - `py2neo` Version 4.1.3 -> `pip install py2neo==4.1.3` (Windows) or `pip install 'py2neo==4.1.3'` (Linux)
- *There have been some issues with different version of pip-packages that are installed with py2neo. It might be necessary to revert to an earlier version*


## XLXS reader

- Run `xlsxreader.py`
- Takes path of xlsx file from the argument 
- This script creates python files containting dicts with the data for the "upper" level and the "simutool" from the input excelfile . I
- These files are saved to `.` where python was run.


## Neo4j Graph Populator

### The python-files containing the information for the domain-models

- They need to have the following structure:

``` python
classes =  [ { node: {property: ... }, ... ]
relations = [ { relation: {property: ... }, ... ]
namespaces = [ { namespace: {property: ... }, ... ]
```

- `classes` is always required.
- `relations` and `namespaces` are optional.
- `relations` holds all object-property relations of this domain-model.
- `namespaces` holds information for the namespaces added by this domain-model/on this level. Therefore, the uppermost domain-model file needs to include a namespaces dict in order to introduce namespaces at all.

### Creating cypher queries

- Run `import_domain_model.py` with the python files storing information about each level as arguments to create the cypher-queries
- The python file in the first argument is expected to contain the information of the uppermost level. This means it requires at least dicts called `classes` and `namespaces`
- This script creates cypher queries for creating the class and namespace nodes and the relations between classes (subclass relations as well as object-property relations).
- It is possible to state a database connection via the `--db` flag. Following syntax is required: `user:pwd@ip:port`. If the `--db` flag is set but the given parameter does not match the expected structure, you will be asked to entered `ip:port`, `user` and `pwd` manually. The stated database is than erased and filled with the information form the python-dict-files in the arguments.
- If no `--db` flag is set the cypher queries will just be printed to std-out.
- All available options and parameters can be seen using the `--help` or `-h` option. The help option will provide the `helpfile.txt`.
- **Caution:** The script always erases the database completely after connecting. This means also entries that are not being altered by this script are lost. Also errors while creating and running the cyphers will erase the databse in order to leave the databas in a consistent state (empty)
- Erros during the import will just exit before connection to the databse  and do therefore not erase the database




CHANGELOG.md
# Changelog


## [Unreleased]
### Added

### Changed

### Removed

### Fixed


## 1.1.0 (2019-04-29)
### Added
- function create_property_nodes, now properties are represented as nodes
- function create_req_property_relations, creates relations :required_property from a class to its required properties
- function create_opt_property_relations, creates relations :optional_property from a class to its optional properties
- dict properties in the upper.py file
### Changed
- 
- 
- 

### Removed
- Node attributes "optional_properties" and "required_properties" because we represent them as relations now

### Fixed


sample-inputs/simutool.py
  # Low-level details of the domain model. are specific to the target domain, and hence will contain specific information about manufacturing processes, resources, and so on. This information can be changed according to the specifications of the end-users 
# Created 2019-07-12 16:49:32.430000 

# Dict storing class info: 
classes = [{'ToolDevelopment': {'admin': '',
                      'description': 'tbd',
                      'html_info': '',
                      'identifier': 'http://example.org/tbox/tooldevelopment',
                      'label': 'TBox',
                      'ontology_level': 'simutool',
                      'optional_property': [],
                      'pl': 'Tool Development Activities',
                      'required_property': [],
                      'sing': 'Tool Development',
                      'subclass_of': 'Activity',
                      'title': 'ToolDevelopment',
                      'version': 'v3.1'}}]


sample-inputs/upper.py
# The upper domain model is generic across different data management domain applications. 
# Created 2019-07-12 16:49:32.398000 

# Dict storing class info: 
classes = [{'KBMSThing': {'admin': 'yes',
                'description': 'Mother of all knowledge instances',
                'html_info': 'Add a new general item that does not have a specific type.',
                'identifier': 'http://example.org/tbox/kbmsthing',
                'label': 'TBox',
                'ontology_level': 'upper',
                'optional_property': [],
                'pl': 'KBMS Things',
                'required_property': ['description', 'identifier', 'title'],
                'sing': 'KBMS Thing',
                'subclass_of': 'NULL',
                'title': 'KBMSThing',
                'version': 'v3.1'}},
 {'Agent': {'admin': 'yes',
            'description': 'Things or people that can create knowledge, like users, organization, or systems',
            'html_info': '',
            'identifier': 'http://example.org/tbox/agent',
            'label': 'TBox',
            'ontology_level': 'upper',
            'optional_property': [],
            'pl': 'Agents',
            'required_property': [],
            'sing': 'Agent',
            'subclass_of': 'KBMSThing',
            'title': 'Agent',
            'version': 'v3.1'}},
 {'Activity': {'admin': '',
               'description': 'A frame for certain knowledge, like a project or a project phase',
               'html_info': '',
               'identifier': 'http://example.org/tbox/activity',
               'label': 'TBox',
               'ontology_level': 'upper',
               'optional_property': [],
               'pl': 'Activities',
               'required_property': [],
               'sing': 'Activity',
               'subclass_of': 'KBMSThing',
               'title': 'Activity',
               'version': 'v3.1'}},
 {'Data': {'admin': '',
           'description': 'A digital representation of knowledge',
           'html_info': '',
           'identifier': 'http://example.org/tbox/data',
           'label': 'TBox',
           'ontology_level': 'upper',
           'optional_property': ['subject'],
           'pl': 'Data',
           'required_property': [],
           'sing': 'Data',
           'subclass_of': 'KBMSThing',
           'title': 'Data',
           'version': 'v3.1'}}]

# Dict storing relation info. 
relations = [{'references': {'description': 'A related KBMSThing  that is referenced, cited, or otherwise pointed to by the described KBMSThing',
                 'from_entity': 'KBMSThing',
                 'identifier': 'http://example.org/tbox/references',
                 'label': 'object_property',
                 'level': 'upper',
                 'namespace': 'dcterms',
                 'to_entity': 'KBMSThing'}},
 {'relation': {'description': 'A relation between two KBMSThings (bi-directional version of references)',
               'from_entity': 'KBMSThing',
               'identifier': 'http://example.org/tbox/relation',
               'label': 'object_property',
               'level': 'upper',
               'namespace': 'dcterms',
               'to_entity': 'KBMSThing'}},
 {'haspart': {'description': 'A related KBMSThing that is included either physically or logically in the described KBMSThing.',
              'from_entity': 'KBMSThing',
              'identifier': 'http://example.org/tbox/haspart',
              'label': 'object_property',
              'level': 'upper',
              'namespace': 'dcterms',
              'to_entity': 'KBMSThing'}},
 {'rights': {'description': 'The Document includes  a statement about various property rights associated with the KBMSThing, including intellectual property rights.',
             'from_entity': 'KBMSThing',
             'identifier': 'http://example.org/tbox/rights',
             'label': 'object_property',
             'level': 'upper',
             'namespace': 'dcterms',
             'to_entity': 'Document'}},
 {'rightsholder': {'description': 'A person or organization owning or managing rights over the KBMSThing',
                   'from_entity': 'KBMSThing',
                   'identifier': 'http://example.org/tbox/rightsholder',
                   'label': 'object_property',
                   'level': 'upper',
                   'namespace': 'dcterms',
                   'to_entity': 'Agent'}}]

# Dict storing property info. 
properties = [
 {'title': {'description': 'A name given to the KBMSThing',
            'identifier': 'http://example.org/tbox/title',
            'label': 'property',
            'label2': 'TBox',
            'namespace': 'dcterms',
            'title': 'title',
            'unique': 'False',
            'xsd_type': 'xsd:string'}},
 {'description': {'description': 'tbd',
                  'identifier': 'http://example.org/tbox/description',
                  'label': 'property',
                  'label2': 'TBox',
                  'namespace': 'dcterms',
                  'title': 'description',
                  'unique': 'False',
                  'xsd_type': 'xsd:string'}},
 {'identifier': {'description': 'A URI that uniquely identifies this thing',
                 'identifier': 'http://example.org/tbox/identifier',
                 'label': 'property',
                 'label2': 'TBox',
                 'namespace': 'dcterms',
                 'title': 'identifier',
                 'unique': 'True',
                 'xsd_type': 'xsd:anyURI'}}]

# Dict storing namespace info. 
namespaces = [{'dcterms': {'comment': '',
              'uri': 'http://purl.org/dc/terms/',
              'url': 'www.dublincore.org/documents/dcmi-terms/'}},
 {'rdfs': {'comment': '',
           'uri': '',
           'url': 'https://www.w3.org/TR/2014/REC-rdf-schema-20140225/'}},
 {'kb': {'comment': 'upper / generic terms for knowledge management system',
         'uri': 'tdb: somewhere under MOBI?',
         'url': ''}},
 {'stkb': {'comment': 'lower / Simutool specific terms',
           'uri': 'tbd: somewhere under MOBI or under simutool.com?',
           'url': ''}},
 {'foaf': {'comment': '', 'uri': '', 'url': 'http://xmlns.com/foaf/spec/'}}]
